### PR TITLE
Issue/31 implement project structure

### LINKED ISSUE #31: Implement project structure
User and Project have a n-m relation,
Project needs:
- project_uuid (pk)
- name
- created_at

user_project:
- project_uuid (pk)
- user_uuid (pk)

Also the blocks have an 1-n relation with the projects:
- block_uuid
- project_uuid (as foreign key)

### PR DESCRIPTION
**Pull request by mottegk**
Add project structure, Closes #31 

Adds models for project and block, adds controller and views for project. 
Error handeling could be expanded and block logic added in the future. Not yet tested.

### CODE REVIEW

**Comment by PaulKalho** [comment]
@mottegk can you draft this again until you moved the files to `workflow-service`?

**Comment by felixevers** [comment]
Please check your linting. @mottegk

**Review by PaulKalho — CHANGES_REQUESTED** [review]
I ignored the missing migrations for now. As I expected, this does not work when we have multiple models. However, this is not bound to this PR. Ignore them for now, I will fix this and then run the migrations for the schema changes (after this PR has been merged) .

You should also definitely think about cascades within our models. I think they can be added within the `relationship` function of `sqlalchemy.orm`

File: core/services/workflow_service/schemas/project.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+class CreateProjectResponse(BaseModel):
+    project_uuid: UUID
+
+
+class ReadProjectRequest(BaseModel):
+    project_uuid: UUID
+
+
+class ReadProjectResponse(Project):
+    pass
```
[the remark this answers is not in the document]
Comment (mottegk) [reply core/services/workflow_service/schemas/project.py:35]: I was unsure if I should use project directly, as then there are some views which have a specific response schema and some just use project, I will change it to just use Project.

File: core/services/workflow_service/models/project.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from utils.database.connection import Base
+
+
+class Project(Base):
+    __tablename__ = "projects"
+
+    uuid = Column(UUID(as_uuid=True), primary_key=True,
+                  default=uuid.uuid4)
+    name = Column(String(100), nullable=False)
+    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
```
[the remark this answers is not in the document]
Comment (mottegk) [reply core/services/workflow_service/models/project.py:17]: I changed the import, thanks for pointing it out to me :)

File: core/services/workflow_service/schemas/project.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+class Project(BaseModel):
+    uuid: UUID
+    name: str
+    created_at: datetime
+    users: List[UUID]
+    blocks: List[UUID]
+
+    class Config:
+        orm_mode = True  # ORM mode: easy conversion from SQLAlchemy objects
```
[the remark this answers is not in the document]
Comment (mottegk) [reply core/services/workflow_service/schemas/project.py:17]: Oh, yeah I'm sorry I used an outdated source, its `from_attributes` now. I changed the comment a bit, but basically it helps convert the sqlalchemy model into a pydantic model and validates the data.

**Comment by mottegk** [comment]
> You should also definitely think about cascades within our models. I think they can be added within the `relationship` function of `sqlalchemy.orm`
@PaulKalho Do you think I should do cascading all, or just specific cascade options?

**Comment by PaulKalho** [comment]
> > You should also definitely think about cascades within our models. I think they can be added within the `relationship` function of `sqlalchemy.orm`
> > @PaulKalho Do you think I should do cascading all, or just specific cascade options?

What I meant was e.g. that if a Project is beeing deleted, all the blocks should also be deleted.

File: core/services/workflow_service/schemas/project.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+from uuid import UUID
+
+from pydantic import BaseModel
+from datetime import datetime
+from typing import List
+
+
+# TODO:Still missing any field validation (empty name, uuid is not UUID..)
```
Comment (PaulKalho) [inline core/services/workflow_service/schemas/project.py:8]: nit: `uuid is not UUID` is actually beeing checked, can be removed from the TODO

File: core/services/workflow_service/controllers/project_controller.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    return project.uuid
+
+
+def read_project(project_uuid: UUID) -> Project:
+    db: Session = next(get_database())
+
+    project = db.query(Project).filter_by(uuid=project_uuid).one_or_none()
+
+    if not project:
+        raise ValueError("Project not found")
```
Comment (PaulKalho) [inline core/services/workflow_service/controllers/project_controller.py:42]: Use `HTTPException`. Please check all occurences of `ValueError`. I think this makes sense in all of the usages (especially within the controller)

File: core/services/workflow_service/models/user_project.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+from uuid import UUID
```
Comment (PaulKalho) [inline core/services/workflow_service/models/user_project.py:1]: Dont use the `uuid` package here. As you are defining a `sqlalchemy` model, you should use: `from sqlalchemy.dialects.postgresql import UUID`.

Use the `uuid` package for the default value.

File: core/services/workflow_service/models/block.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+from uuid import UUID
```
Comment (PaulKalho) [inline core/services/workflow_service/models/block.py:1]: also change this import to the `sqlalchemy` one

File: core/services/workflow_service/models/project.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+from uuid import UUID
```
Comment (PaulKalho) [inline core/services/workflow_service/models/project.py:1]: and also this one :)

File: core/services/workflow_service/models/project.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+class Project(Base):
+    __tablename__ = "projects"
+
+    uuid = Column(UUID(as_uuid=True), primary_key=True,
+                  default=uuid.uuid4)
+    name = Column(String(100), nullable=False)
+    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
+
+    users = relationship("User", secondary="user_project",
+                         back_populates="projects")
```
Comment (PaulKalho) [inline core/services/workflow_service/models/project.py:20]: I am not 100% sure. But I believe `back_populate` here only works if you also add a `projects = relationship(..)` to the `user` model. 

What do you think? This is what I found from the docs:

`relationship()` Provide a relationship between *two* mapped classes.
`back_populates` Indicates the name of a [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship) on the related class that will be *synchronized* with this one. It is usually expected that the [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship) on the related class also refer to this one. (...)

**Review by PaulKalho — CHANGES_REQUESTED** [review]

File: core/services/workflow_service/models/project.py (hunk 9 above)
Comment (mottegk) [reply core/services/workflow_service/models/project.py:20]: Yes, you're right, I forgot, that I didn't put a projects relationship in User. I'll either remove the back_populates or add the relationship.

**Review by PaulKalho — APPROVED** [review]

**Pull request merged**